### PR TITLE
Optimize HSTU qlen=1 causal and local attention on SM100, SM103, and SM107

### PR DESCRIPTION
**Pull request by cb521**
## Before submitting

- [ ] I agree to license this contribution under the terms of [LICENSE.txt](https://github.com/NVIDIA/cudnn-frontend/blob/develop/LICENSE.txt).
- [x] I ran `pre-commit run` and committed any formatting changes.
- [x] I reviewed the Hard Rules in the `AGENTS.md` files for the directories this PR touches; the optimized execute paths avoid device-to-host reads, use symbolic packed extents, and retain launch-stream ordering.
- [ ] Requested labels: `cat-enhancements`, `mod-cutedsl`, `area:sparse_attention`, and `orig-nv-eng`. GitHub does not allow the fork author to apply labels to the upstream repository.

## Affected area

- FE OSS kernels or CuTeDSL
- Benchmarks or performance

## Summary

- Specialize BF16 HSTU `qlen=1` forward for causal and bounded-local masks on SM100, SM103, and SM107 with an M64/N128 tensor-core kernel.
- Use only two measured forward configurations in production dispatch: one for D64/D128 and one for D256.
- Keep experimental forward candidate names and their knob translation in the standalone benchmark instead of the production dispatcher.
- Add a direct CUDA-core backward kernel for D64/D128/D256 and a metadata-only host selector between unsplit execution, split8 on SM100/SM103, and split13 on SM107.
- Keep unsplit D256 backward blocks at 256 or more threads so every dQ element is written.
- Use symbolic packed-Q, packed-KV, and batch extents so compiled artifacts can be reused across continuous-batching shapes.

## Why

The general HSTU path performs substantial unnecessary work for a single right-aligned query, especially with a bounded local window. The specialized path limits forward math to attended KV blocks and avoids backward arithmetic outside the window while still explicitly writing zero dK/dV rows required by the API.

Selection remains host-only: it does not synchronize the device, inspect tensor values, benchmark at runtime, or use a per-shape lookup table. Forward candidate benchmarking is a development-time activity; normal execution directly selects one of the two fixed configurations from head dimension.

## SM100 and SM103 paired tuning

SM100 uses the same SM10x kernel implementation and automatic schedule as SM103: the M64 forward path, split8 backward default, and matching block-size thresholds. SM107 keeps its independently measured split13 policy and Rubin-only two-CTA path.

This reuse decision was validated with paired sweeps on a 1000 W NVIDIA B200 (SM100) and a 1100 W NVIDIA B300 (SM103), using the same CUDA 13.0, PyTorch 2.13, CuTeDSL environment, workload shapes, and measurement settings. Each GPU ran 51 broad and 51 focused benchmark configurations spanning D64/D128/D256, causal and local masks, average KV lengths from 128 to 4096, batch sizes from 16 to 1024, forward split candidates, backward split candidates, and 128/256/384/512-thread schedules.

The two GPUs produced the same meaningful kernel-family, split, and thread crossovers. The few different nominal winners were sub-percent launch-limited ties with overlapping sample ranges, not stable SM100-specific behavior. Therefore this PR intentionally keeps one SM100/SM103 policy instead of adding an architecture-specific lookup table.

## API and compatibility impact

No public API signature changes. The automatic optimized path applies to BF16 `qlen=1`, causal or bounded-local masking, D64/D128/D256, matching Q/K/V heads, non-paged KV, and supported direct layouts on SM100/SM103/SM107. FP16 and unsupported mask/layout combinations retain the existing fallback behavior. D32, newly supported by `develop`, remains on the general HSTU path.

## Benchmark

`benchmark/hstu_attention/benchmark_hstu_qlen1.py` is the only added benchmark file. It owns the experimental forward candidate list and translates candidate names into low-level kernel knobs. It covers causal, bounded-local, and full masks; forward, backward, or both directions; correctness checks plus timed runs; and automatic dispatch versus explicit implementation comparisons.

Historical B300/Rubin runs showed 1.17x–1.58x forward and 2.07x–2.50x backward improvements for the representative D128 causal workload (H=4, average KV about 2048). The B200 measurements validate parameter selection; they were not a baseline-versus-optimized run, so this PR does not claim a separate B200 speedup number.

## Testing

- Rebased onto current `upstream/develop` (`14847e8`).
- `pre-commit run --files <changed files>` — passed after the rebase and dispatch cleanup.
- `python -m py_compile <changed Python files>` — passed.
- `git diff --check upstream/develop` — passed.
- Post-cleanup B200 and B300 forward checks for D32/D64/D128/D256 — passed against the PyTorch reference; D32 remains on the general path.
- A benchmark-only `tc-m64-warp1-split2` candidate — compiled and passed correctness on B200, confirming the tuning entry remains usable outside production dispatch.
- B200 and B300 automatic forward/backward smoke tests — passed.
- B200 and B300 causal correctness for automatic D128/D256 plus explicit leading D64/D128/D256 split candidates — passed.
- B200 and B300 bounded-local automatic correctness for D64/D128/D256 — passed.
- B200 and B300 paired broad, focused, and short-KV thread sweeps — completed with matching architecture trends.
- D256 unsplit regression: the old 128-thread schedule left all 524,288 upper-half dQ elements unwritten; the fixed 256-thread schedule writes the full tensor and matches the PyTorch reference on both B200 and B300 (maximum absolute error about `1.0e-5`).
- Short-KV bounded-local benchmark with a requested window larger than `max_seqlen_k` — passed on B200 and B300 after clamping the effective benchmark window.
- Local pytest execution remains unavailable in this checkout because it has no built `cudnn._compiled_module`; GPU validation used the cluster environment described above.

## Summary by CodeRabbit

- **New Features**
  - Added support for HSTU attention workloads with one query token per sequence.
  - Added optimized forward and backward execution for variable key/value lengths, local attention windows, and supported data types.
  - Added automatic workload- and hardware-aware execution selection, with optional algorithm controls.
  - Added configurable CUDA benchmarking with correctness checks, timing results, and JSON output.
  - Added execution caching for repeated workloads across runtime shapes and head dimensions.

- **Bug Fixes**
  - Improved split key/value execution and gradient accumulation for single-query workloads.
  - Improved masking for local attention windows and boundary conditions.

### CODE REVIEW

File: python/cudnn/hstu_attention/_interface.py [hunk 1]
```diff
@@ -158,6 +186,286 @@ def _supports_bwd_direct_grad_layout(t: torch.Tensor) -> bool:
...
 
 
+def _select_q1_bwd_num_threads(capability: tuple[int, int], batch_size: int, num_heads: int, split_kv: int) -> int:
+    """Select the block size paired with the qlen=1 backward split."""
+    base_ctas = batch_size * num_heads
+    if split_kv == 1 and base_ctas >= 2048:
+        # Once the unsplit grid already has enough CTAs, a small block avoids
+        # spending 12--16 warps on each short sequence. Split kernels already
+        # reach the same four-warp floor below.
+        return 128
```
[the remark this answers is not in the document]
Comment (cb521) [reply python/cudnn/hstu_attention/_interface.py]: Fixed in d5f6019. The selector now receives head_dim and returns at least 256 threads for unsplit D256; the kernel constructor also asserts the dQ coverage invariant. I reproduced the old failure on B200 (all 524,288 upper-half dQ elements remained unwritten), then verified the fix against a PyTorch reference on both B200/SM100 and B300/SM103 (max abs error about 1.0e-5).

File: python/cudnn/hstu_attention/_interface.py [hunk 2]
```diff
@@ -230,6 +627,47 @@ def hstu_varlen_fwd_100(
...
+        "tc-m64-16dp-tail-kv5",
+        "tc-m64-16dp-tail-kv5-split2",
+        "tc-m64-16dp-tail-kv5-split4",
+        "tc-epi1",
+        "tc-epi1-split2",
+        "tc-epi1-split4",
+    ):
+        raise ValueError(f"Unsupported qlen=1 forward algorithm: {_q1_fwd_algorithm}")
+    if q1_split_kv > 1 and not q1_split_supported:
+        raise ValueError("The split-KV qlen=1 forward algorithms require causal BF16 D=128 qlen=1 with matching Q/K/V heads")
```
[the remark this answers is not in the document]
Comment (cb521) [reply python/cudnn/hstu_attention/_interface.py]: Fixed in d5f6019. Both messages now match the actual D64/D128/D256 support predicate.

File: test/python/fe_api/hstu_attention/test_hstu_attention.py [hunk 3]
```diff
@@ -1501,6 +1511,389 @@ def test_varlen_tail_and_asymmetric_lengths_match_pytorch(head_dim):
...
+            alpha=0.7,
+            scaling_seqlen=256.0,
+            causal=False,
+            window_size=(63, 0),
+        )
+        torch.testing.assert_close(actual.cpu().float(), expected, rtol=4e-2, atol=4e-2)
+
+    assert len(_interface.hstu_varlen_fwd_100.compile_cache) == 1
+    key = next(iter(_interface.hstu_varlen_fwd_100.compile_cache))
+    assert key[14] == 1
```
[the remark this answers is not in the document]
Comment (cb521) [reply test/python/fe_api/hstu_attention/test_hstu_attention.py:1874]: Fixed in d5f6019. The test now asserts compile-key index 13 for split-KV and keeps index 14 as the independent head-count check.

**Review by vedaanta — APPROVED** [review]

**Comment by vedaanta** [comment]
@cudnn-ci-bot run frost

**Comment by vedaanta** [comment]
@cudnn-ci-bot run oss

File: python/cudnn/hstu_attention/_interface.py [hunk 4]
```diff
@@ -412,7 +829,63 @@ def hstu_varlen_bwd_100(
...
+    elif q1_direct_supported:
+        selected_q1_bwd_algorithm = _select_q1_bwd_algorithm(_q1_bwd_algorithm, batch_size, q.device)
+    else:
+        selected_q1_bwd_algorithm = _q1_bwd_algorithm
+    use_q1_direct_kernel = q1_direct_supported and selected_q1_bwd_algorithm in _Q1_BWD_DIRECT_SPLITS
+    # Keep every lane on one aligned 128-bit vector. A warp consequently packs
+    # four D64 rows, two D128 rows, or one D256 row.
+    q1_rows_per_warp = 256 // head_dim if selected_q1_bwd_algorithm.startswith("direct-pair") else 1
+    use_q_major_scheduler = selected_q1_bwd_algorithm in ("tc", "tc-small")
+    use_q1_small_mma = selected_q1_bwd_algorithm == "tc-small"
```
[the remark this answers is not in the document]
Comment (cb521) [reply python/cudnn/hstu_attention/_interface.py:870]: Fixed in df2ccfd. Explicit D256 tc/tc-small tuning requests now raise before the dedicated D256 fallback can run, and an L0 regression test covers both labels. The guard checks the explicit private tuning request (rather than an auto-resolved fallback) so the public auto path remains unchanged. The regression test failed for both parameters before the guard and passes for both after it.

**Review by jiayus-nvidia — APPROVED** [review]
